packages/@expo/cli/src/prebuild/resolveTemplate.ts

```diff
@@ -91,21 +93,31 @@ function hasRepo({ username, name, branch, filePath }: RepoInfo) {
 }
 
 async function downloadAndExtractRepoAsync(
-  root: string,
-  { username, name, branch, filePath }: RepoInfo
+  { username, name, branch, filePath }: RepoInfo,
+  props: ExtractProps
 ): Promise<string> {
-  const projectName = path.basename(root);
-
-  const strip = filePath ? filePath.split('/').length + 1 : 1;
-
   const url = `https://codeload.github.com/${username}/${name}/tar.gz/${branch}`;
+
   debug('Downloading tarball from:', url);
-  return await extractNpmTarballFromUrlAsync(url, {
-    cwd: root,
-    name: projectName,
-    strip,
-    fileList: [`${name}-${branch}${filePath ? `/${filePath}` : ''}`],
+
+  // Extract the (sub)directory into non-empty path segments
+  const directory = filePath.replace(/^\//, '').split('/').filter(Boolean);
+  // Remove the (sub)directory paths, and the root folder added by GitHub
+  const strip = directory.length + 1;
+  // Only extract the relevant (sub)directories, ignoring irrelevant files
+  const filter = createGlobFilter(!directory.length ? '*/**' : `*/${directory.join('/')}/**`, {
+    // Enable dotfiles for `.xcode.env`
+    dot: true,
+    // Keep this up to date with the `.npmignore` files in expo/expo/templates/<template>/.npmignore
+    ignore: [
+      '**/.gitattributes',
+      '**/.npmignore',
+      '**/ios/HelloWorld.xcworkspace/**',
+      '**/ios/.xcode.env.local',
+    ],
```

Comment:
Let me try if we can add inverse ignores. If we keep `dot: false` (default), and explicitly add `!**/ios/.xcode.env` as ignore, we could get rid of some options here.

Reply:
Negative ignores breaks picomatch, but we can provide multiple globs in the filter.